slick-hikaricp/src/main/scala/slick/jdbc/hikaricp/HikariCPJdbcDataSource.scala

```diff
@@ -1,16 +1,66 @@
 package slick.jdbc.hikaricp
 
-import java.sql.{Driver, Connection}
+import java.sql.{Connection, Driver}
+import java.util.concurrent.atomic.AtomicInteger
+import java.util.concurrent.locks.ReentrantLock
+
 import com.typesafe.config.Config
 import slick.SlickException
-import slick.jdbc.{JdbcDataSourceFactory, JdbcDataSource}
+import slick.jdbc.{JdbcDataSource, JdbcDataSourceFactory}
 import slick.util.ConfigExtensionMethods._
+import slick.util.Logging
 
 /** A JdbcDataSource for a HikariCP connection pool.
   * See `slick.jdbc.JdbcBackend#Database.forConfig` for documentation on the config parameters. */
-class HikariCPJdbcDataSource(val ds: com.zaxxer.hikari.HikariDataSource, val hconf: com.zaxxer.hikari.HikariConfig) extends JdbcDataSource {
-  def createConnection(): Connection = ds.getConnection()
+class HikariCPJdbcDataSource(val ds: com.zaxxer.hikari.HikariDataSource, val hconf: com.zaxxer.hikari.HikariConfig)
+  extends JdbcDataSource
+  with Logging {
+
+  private[this] val poolLock: ReentrantLock = new ReentrantLock()
+  private[this] val condition = poolLock.newCondition()
+  private[this] var count: Int = 0
+
+  def createConnection(): Connection = {
+    poolLock.lock()
+    try {
+      if (logger.isDebugEnabled) logger.debug("number of connections in use = "+count)
+      while (count == hconf.getMaximumPoolSize) {
+          condition.await()
+      }
+      val c = new DelegateConnection(ds.getConnection()) {
+        override def close(): Unit = {
+          closeConnection(conn)
+        }
+      }
+      count += 1
+      if (count == hconf.getMaximumPoolSize) {
+        logger.info("Maximum number of connections in use reached, pausing")
+        pause()
+      }
+      c
+    } finally {
+      poolLock.unlock()
+    }
+  }
+
+  private[this] def closeConnection(connection: Connection) = {
+    poolLock.lock()
+    try {
+      connection.close()
+    } finally {
+      count -= 1
+      if (count == hconf.getMaximumPoolSize -1) {
+        logger.info("Connection freed below maximum, resuming")
+        condition.signal()
+        resume()
+      }
+      poolLock.unlock()
+    }
+  }
+
+
   def close(): Unit = ds.close()
```

Comment:
This should be fine.  HikariCP will attempt to wait, if I remember correctly 5 seconds, for in-use connections to return to the pool.  New callers to getConnection() are kicked out with a SQLException.  After the 5 seconds, HikariCP starts  aborting connections (Connection.abort()) that are still in-use.

When a pool is shutdown when connections are in use, there is never a truly graceful way -- how long to you wait, etc.  You certainly don't want the app to hang for minutes waiting for long running queries to end, or god-forbid a blocked thread which might never exit.  Hopefully, the application itself is in charge of shutting down the datasource, and has properly terminated (or waited for) it's threads.

Anyway, close is thread safe, so as far as HikariCP is concerned there is nothing else that needs to be done here.  But if slick wants to be even more "polite" to executing or queued threads before calling close() it certainly is within your control to do so.